engine/common/splitter/engine.go

```diff
@@ -0,0 +1,149 @@
+package splitter
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/network"
+	"github.com/rs/zerolog"
+)
+
+type Engine struct {
+	mu      sync.Mutex
+	unit    *engine.Unit               // used to manage concurrency & shutdown
+	log     zerolog.Logger             // used to log relevant actions with context
+	engines map[module.Engine]struct{} // stores registered engines
+	channel network.Channel            // the channel that this splitter listens on
+}
+
+func New(
+	log zerolog.Logger,
+	channel network.Channel,
+) *Engine {
+	e := &Engine{
+		unit:    engine.NewUnit(),
+		log:     log.With().Str("engine", "splitter").Logger(),
+		engines: make(map[module.Engine]struct{}),
+		channel: channel,
+	}
+
+	return e
+}
+
+func (e *Engine) RegisterEngine(engine module.Engine) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if _, ok := e.engines[engine]; ok {
+		return errors.New("engine already registered with splitter")
+	}
+
+	e.engines[engine] = struct{}{}
+
+	return nil
+}
+
+func (e *Engine) UnregisterEngine(engine module.Engine) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	delete(e.engines, engine)
+}
+
+// Ready returns a ready channel that is closed once the engine has fully
+// started. For the splitter engine, this is true once all of the
+// registered engines have started.
+func (e *Engine) Ready() <-chan struct{} {
+	return e.unit.Ready(func() {
+		for engine := range e.engines {
```

Comment:
When this is engine is initially created by the Splitter Network, the `engines` map will be empty. The splitter network also does not call the `Ready()` function. Hence we probably don't need to do anything in `Ready`.

Reply:
`Ready()` is called by `Run()` in `scaffold.go` [here](https://github.com/onflow/flow-go/blob/master/cmd/scaffold.go#L746) and [here](https://github.com/onflow/flow-go/blob/master/cmd/scaffold.go#L605) to make sure all components are ready before starting the node. Since `Run()` is only called [at the very end](https://github.com/onflow/flow-go/blob/master/cmd/access/main.go#L344) of flow node setup, the `engines` map will already be initialized by then.

It is true that if when we use the splitter network, the `Ready()` function will not be called since the splitter network creates the splitter engine instances on its own. However, I think it's conceivable that there may be cases where someone would want to use the splitter engine on its own by directly registering it with the regular network, in which case `Ready()` and `Done()` is necessary

Reply:
this _may_ create a circular loop. This `Ready()` depends on the `Ready()` of other engines. But the first thing an engine does is register with a network. In this case that would be the splitter network. And then we may have a deadlock engineX.Ready() -> invokes network.Register -> which invokes Splitter engine. Scaffold then does a splitter engine.Ready() which waits on engineX.Ready().
Its not possible right now but wanted to bring it to ur attention.

Reply:
> And then we may have a deadlock engineX.Ready() -> invokes network.Register -> which invokes Splitter engine. 

Currently, engines will register with the network in the constructor, not in the `Ready()` method. Which makes sense, given than in most scenarios a failure to register with the network is an error that would warrant failing the instantiation of the engine altogether. 

As a side note, the possibility that an engine *could* contain the antipattern just described is further justification for not passing in the network instance directly to engines, and instead passing in conduits: dapperlabs/flow-go#5124 

Even if some engine did call `network.Register` inside its `Ready` method, this would not cause a deadlock on its own. The only way a deadlock could occur is if an engine registered with the splitter engine, *AND* its `Ready` method called the splitter engine's `Ready` method. 

I think we could perhaps add comments / documentation to highlight this pitfall, but I'm not sure that this warrants completely getting rid of the splitter engine's `Ready` method.